cfpb/consumerfinance.gov
# Automatically import FIG data points from JSON

Upon save, preview, or publish, a FIGContentPage now downloads JSON (following a specific structure representing FIG data points) from a specified URL, parses, and saves it to the page. The content editor places a FigDataPointsBlock in whatever Section of the page they want the data points to appear in.

On page load, the data point info is placed into the data points template in the section specified by the content editor. Links to the data points are added to the table of contents.


## Additions

- Data models (with migrations), an import script, a jinja template, to support the new ability to import structured JSON and have it inserted into a page

## Changes

- Minor changes to things like FIG section numbering to support the change

## How to test this PR

1. Running this branch, in Wagtail, create a Fig Content Page
2. In a Fig Section in the content of the page, add a Fig Data Points block
3. Find or create a structured JSON file of FIG data points. An example is at GHE/raw/CFPB/1071-fig-content/blob/master/test_data_points.json
4. In the Data Points tab of the Wagtail editor, enter the download URL of the JSON file
5. Preview or publish the page. 
6. See that the data points are displayed in basically the format we're going for
7. See that the table of contents correctly links to the data points and the anchor links are sensible


## Notes and todos

- ~~Ensure that the data points block is only inserted at the Section level, since that's the only way the formatting and section numbering make sense.~~
- Also, find a nicer way to validate the JSON schema. Consider using Wagtail Snippets to hold the JSON data, rather than downloading from a separate location.
- Document how to use this feature in Hubcap


## Checklist



  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance
